service/storage_service.cc

```diff
@@ -6080,6 +6171,9 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
     ser::storage_service_rpc_verbs::register_tablet_stream_data(&_messaging.local(), [this] (locator::global_tablet_id tablet) {
         return stream_tablet(tablet);
     });
+    ser::storage_service_rpc_verbs::register_tablet_cleanup(&_messaging.local(), [this] (locator::global_tablet_id tablet) {
```

Comment:
Why not have single tablet RPC that will multiplex to different functionality. Boilerplate code (like barriers) can be reused.

Reply:
No multiplex is terrible. Pretty soon you have lots of parameters and you aren't sure which are used.

Reply:
If you send `std::variant<param_for_stream, param_for_cleanup>` you both multiplex and know exactly which parameters are used for what. OTOH if you constantly add high level rpc verbs in the end you are not sure which set is exactly responsible for what.

Reply:
> you are not sure which set is exactly responsible for what.

I don't understand this point, say more.

Reuse can be achieved in the implementation by template methods or composition.

Reply:
> > you are not sure which set is exactly responsible for what.
> 
> I don't understand this point, say more.
> 

If we have all RPCs related to tablets in a single RPC and multiplex internally it is easy to see what rpcs are tablets related even when you add more RPCs later. If you allocate new global RPC each time you add one in then end we will have a set of verbs in a global verb space and it will be harder to see which set of RPCs are responsible to tables. We multiplex in vnode topology and node ops.

> Reuse can be achieved in the implementation by template methods or composition.

Yes, it can. But multiplexing has other advantages as well.

Reply:
All tablet verbs have "tablet" in the name, so I would say you can see which ones are tablet specific. We already do that with REPAIR_*, PAXOS_*, RAFT_*, etc.

Reply:
Those others I am familiar with do not have common boilerplate. Additional complication with PAXOS that some RPCs parameters are dictated by the way the generic code use them (PAXOS_LEARN is used by generic mutation sending code). Raft one have different semantics (some are one way some two ways). But I agree that we have both examples in the code. Here I advocate for multiplexing not because the RPCs belong to the same subsystem, but because they are used in a similar way and have similar semantics. They both are issued by the tablet coordinator as part of tablet movement and expect to sync with raft sate and check for outdated coordinator. If we will have another tablet related RPC that has different semantics then it should be a separate verb.

Of course you still can have the same boilerplate without duplicating the code by calling the same function from different RPC, but then what advantage does it have over multiplexing?

Reply:
Multiplexing is just bad. Yes you can have a function called do_something(variant<x, y,z> parameters), but is that preferable to do_x(x_parameters), do_y(y_parameters), do_z(z_parameters)?

Related: https://lwn.net/Articles/858332/

Reply:
Sure, but I don't think tablet streaming and tablet cleanup have enough in common here. At least that's how it looks from here, on the high ground.